2021/TurcoZiccolella/app.py

```diff
@@ -0,0 +1,491 @@
+from utils.talker import Talker
+from utils.pipeline import Pipeline
+from utils.init_functions import Setup
+from utils.grafana_dashboard import DashBoard
+from nfstream.streamer import NFStreamer
+
+import logging
+import subprocess
+import signal
+import sys
+import tempfile
+import os
+import datetime
+import traceback
+import rrdtool
+import ipaddress
+import operator
+import configparser
+from concurrent.futures.thread import ThreadPoolExecutor
+import multiprocessing
+import time
+
+# Formato di display della data
+p = '%Y-%m-%d--%H:%M:%S'
+pid_child = 0
+
+def producer(pl, chosen_if, rrd_step):
+    try:
+        print("producer is active")
+        p_out,p_in = multiprocessing.Pipe()
+
+        global pid_child
+        pid_child = os.fork()
+
+        if (pid_child == 0):#CHILD
+
+            def signal_handler(sig, frame):
+                    print('Stopping TcpDump')
+                    process.terminate()
+                    sys.exit(0)
+            signal.signal(signal.SIGINT, signal_handler)
+
+            os.setuid(0)#Always enabled root
+            while True:
+                pcapfile,path = tempfile.mkstemp(suffix='.pcap')  # creates the file with the umask 600, the owner of the file may read from and write to it
+                process = subprocess.Popen(['timeout', rrd_step, 'tcpdump', '-i', chosen_if,'-w', path, '-U'])
+                process.wait()
+                p_in.send(path)
+
+        else:#FATHER
+
+            while True:
+
+                path = p_out.recv() #Bloking call
+                finish_date = datetime.datetime.now()
+                pl.set_message(finish_date.timestamp(), path)
+            
```

Comment:
Processo figlio che esegue la cattura con tcpdump e con permessi di root, il file è creato con mkstemp e salvato quindi in una cartella del sistema operativo che contiene file temporanei, è accedibile solamente dall'owner del file (quindi root)

Reply:
Le capabilities sono più dinamiche dell'uid , però non posso essere assegnate ad uno script. L'unico modo per assegnarle è per ereditá (garantita dalla fork) dall'interterprete.

L'euid a sua volta è assegnato al processo e  una chiamata seteuid manda dei segnali ai vari thread del processo per modificare il proprio euid (variabile unica per thread ma resa coerente tramite segnali). 

Quindi il miglior metodo per evitare  problemi di concorrenza era eseguire una fork (usare una lock sarebbe stato inefficiente poichè tutto il programma avrebbe mantenuto i permessi di root durante la cattura) , in modo tale che io possa cambiare l'euid solo di una parte di codice incapsulata in un nuovo processo.

La comunicazione tra processo padre e processo figlio è implementata tramite una pipeline ed in fase di chiusura (SIGINT) viene inviato un segnale al processo figlio affinchè termini la sua esecuzione.